### ISSUE TITLE
The nature of Dirac's delta in SymPy

### ISSUE THREAD

**Issue by Roland-Sands** [sympy/sympy#14098]
In SymPy I have seen expressions involving Dirac's delta being evaluated at a point

In SymPy, is Dirac's delta defined to be a distribution or some other kind of generalized function (so that it cannot be evaluated at points), is it defined to be a hyper-real valued function, or is some other approach being taken

**Comment by asmeurer** [comment]
DiracDelta is a symbolic convenience. Mathematically, δ(x) only makes sense inside of an integral of a certain type, but we allow using it generally as any other expression as a convenience, with the caveat that if you create an expression that is mathematically meaningless, you may get a result that is equally meaningless (garbage in, garbage out). 

We could have just made a DeltaIntegral object which represented `Integral(DiracDelta(x - a)*expr, (x, -oo, oo))`, but then it would be more annoying to work with (for instance, it would not be possible to take `diff(DiracDelta(x)*x, x)`, which is a perfectly reasonable thing to compute assuming you put it inside an integral when you are done).

We also allow evaluating DiracDelta at points. This is perhaps the most dubious, but it fits the intuitive definition of δ(x), and would not be meaningful otherwise. I believe it may make things that would otherwise require special-casing just work (but I could be wrong about that).

**Comment by Roland-Sands** [comment]
Should I close this

**Comment by 2torus** [comment]
@asmeurer I just wrote a similar email in the google group. 
How hard would it be to insert a class ```GeneralizedFunction``` to be a super class of ```Function``` to allow some operations (differentiation, integration, multiplication by a function) but not others like multiplying two objects of ```GeneralizedFunction```?

**Comment by jksuom** [comment]
I don't think it would be too hard to define a `GeneralizedFunction` class, but that should probably not be a superclass of `Function` (nor a subclass). By default, SymPy functions are defined and analytic in the complex plane (apart from some singularities) while the domain of `DiracDelta` and other 'singularity functions' is the real line. The methods of generalized functions would mostly be different from those of ordinary functions and would have to be rewritten.

**Comment by 2torus** [comment]
@jksuom You are right, it should be a separate class. 
Let me see if I can put together a pull request.

**Comment by Roland-Sands** [comment]
@2torus: In the Google groups thread, you wrote: "What was the reasoning to define DiracDelta as a function"
I know 1 reason
Scientists and engineers use Dirac's Delta in a non-rigourous way to create Green's functions
